fix(orc8r): add reindexing of gateway subscriber state to subscriberdb indexer

alexzurbonsen opened the pull request:
## Summary

This PR adds capabilities to subscriberdb indexer to write state of type `gateway_subscriber_state` into the `gateway_subscriber_states` table created in #12949. The logic implemented there splits up the state into its entries and renews the entire state with every indexing run (old entries are deleted and new entries are written into `gateway_subscriber_states` table).

The second commit adds annotations such that the indexer is triggered for this type as well in docker and k8s deployments. 

This is the follow up PR to #12949 and implements step 3 of the plan outlined in [#8621 - comment](https://github.com/magma/magma/issues/8621#issuecomment-1144762649).


## Test Plan

- [x] Unit tests of subscriberdb: indexer and subscriber storage
- [x] manual test of orc8r deployed locally with docker compose

Manual testing is done with grpcurl and the following commands.

1) Send gateway state with three IMSIs.

```
grpcurl -vv \
    -insecure \
    -key /tmp/magma_protos/gateway.key \
    -cert /tmp/magma_protos/gateway.crt \
    -authority state-controller.magma.test \
    -protoset /tmp/magma_protos/out.protoset \
    -d '{"states": [{"type": "gateway_subscriber_state",
"deviceID": "d02a138b-04d3-4710-95aa-d810efc61fc1", "value":"eyJzdWJzY3JpYmVycyI6eyJJTVNJMDAyMDIwMDAwMDAwMTIzIjp7Im1hZ21hLmlwdjQiOlt7ImFjdGl2ZV9kdXJhdGlvbl9zZWMiOjAsImFjdGl2ZV9wb2xpY3lfcnVsZXMiOltdLCJhcG4iOiJtYWdtYS5pcHY0IiwiaXB2NCI6IjE5Mi4xNjguMTI4LjEyIiwibGlmZWN5Y2xlX3N0YXRlIjoiU0VTU0lPTl9SRUxFQVNFRCIsIm1zaXNkbiI6IiIsInNlc3Npb25faWQiOiJJTVNJMDAyMDIwMDAwMDAwMTIzLTEyMzQiLCJzZXNzaW9uX3N0YXJ0X3RpbWUiOjE2NTM0ODQxNDR9XX0sIklNU0kwMDIwMjAwMDAwMDA0NTYiOnsibWFnbWEuaXB2NCI6W3siYWN0aXZlX2R1cmF0aW9uX3NlYyI6MSwiYWN0aXZlX3BvbGljeV9ydWxlcyI6W10sImFwbiI6Im1hZ21hLmlwdjQiLCJpcHY0IjoiMTkyLjE2OC4xMjguMTIiLCJsaWZlY3ljbGVfc3RhdGUiOiJTRVNTSU9OX1JFTEVBU0VEIiwibXNpc2RuIjoiIiwic2Vzc2lvbl9pZCI6IklNU0kwMDIwMjAwMDAwMDA0NTYtMTIzNCIsInNlc3Npb25fc3RhcnRfdGltZSI6MTY1MzQ4NDE0NH1dfSwiSU1TSTAwMjAyMDAwMDAwMDc4OSI6eyJtYWdtYS5pcHY0IjpbeyJhY3RpdmVfZHVyYXRpb25fc2VjIjoyLCJhY3RpdmVfcG9saWN5X3J1bGVzIjpbXSwiYXBuIjoibWFnbWEuaXB2NCIsImlwdjQiOiIxOTIuMTY4LjEyOC4xMiIsImxpZmVjeWNsZV9zdGF0ZSI6IlNFU1NJT05fUkVMRUFTRUQiLCJtc2lzZG4iOiIiLCJzZXNzaW9uX2lkIjoiSU1TSTAwMjAyMDAwMDAwMDc4OS0xMjM0Iiwic2Vzc2lvbl9zdGFydF90aW1lIjoxNjUzNDg0MTQ0fV19fX0=", "version":"0"}]}' \
    localhost:7443 \
    magma.orc8r.StateService/ReportStates
```

The state is written to `states` table and the indexer writes its entries to the `gateway_subscriber_states` table. 

![Screenshot 2022-06-21 at 20 11 09](https://user-images.githubusercontent.com/82914459/174871189-d76aef4b-c784-45d3-bb03-13d8d9d10826.png)

2) After sending an empty state

```
grpcurl -vv \
    -insecure \
    -key /tmp/magma_protos/gateway.key \
    -cert /tmp/magma_protos/gateway.crt \
    -authority state-controller.magma.test \
    -protoset /tmp/magma_protos/out.protoset \
    -d '{"states": [{"type": "gateway_subscriber_state",
"deviceID": "d02a138b-04d3-4710-95aa-d810efc61fc1", "value":"eyJzdWJzY3JpYmVycyI6e319", "version":"0"}]}' \
    localhost:7443 \
    magma.orc8r.StateService/ReportStates
```
the `gateway_subscriber_states` table is indeed empty.

3) Sending mixed states like

```
'{"states": [{"type": "gateway_subscriber_state",
"deviceID": "d02a138b-04d3-4710-95aa-d810efc61fc1", "value":"eyJzdWJzY3JpYmVycyI6eyJJTVNJMDAyMDIwMDAwMDAwMTIzIjp7Im1hZ21hLmlwdjQiOlt7ImFjdGl2ZV9kdXJhdGlvbl9zZWMiOjAsImFjdGl2ZV9wb2xpY3lfcnVsZXMiOltdLCJhcG4iOiJtYWdtYS5pcHY0IiwiaXB2NCI6IjE5Mi4xNjguMTI4LjEyIiwibGlmZWN5Y2xlX3N0YXRlIjoiU0VTU0lPTl9SRUxFQVNFRCIsIm1zaXNkbiI6IiIsInNlc3Npb25faWQiOiJJTVNJMDAyMDIwMDAwMDAwMTIzLTEyMzQiLCJzZXNzaW9uX3N0YXJ0X3RpbWUiOjE2NTM0ODQxNDR9XX0sIklNU0kwMDIwMjAwMDAwMDA0NTYiOnsibWFnbWEuaXB2NCI6W3siYWN0aXZlX2R1cmF0aW9uX3NlYyI6MSwiYWN0aXZlX3BvbGljeV9ydWxlcyI6W10sImFwbiI6Im1hZ21hLmlwdjQiLCJpcHY0IjoiMTkyLjE2OC4xMjguMTIiLCJsaWZlY3ljbGVfc3RhdGUiOiJTRVNTSU9OX1JFTEVBU0VEIiwibXNpc2RuIjoiIiwic2Vzc2lvbl9pZCI6IklNU0kwMDIwMjAwMDAwMDA0NTYtMTIzNCIsInNlc3Npb25fc3RhcnRfdGltZSI6MTY1MzQ4NDE0NH1dfSwiSU1TSTAwMjAyMDAwMDAwMDc4OSI6eyJtYWdtYS5pcHY0IjpbeyJhY3RpdmVfZHVyYXRpb25fc2VjIjoyLCJhY3RpdmVfcG9saWN5X3J1bGVzIjpbXSwiYXBuIjoibWFnbWEuaXB2NCIsImlwdjQiOiIxOTIuMTY4LjEyOC4xMiIsImxpZmVjeWNsZV9zdGF0ZSI6IlNFU1NJT05fUkVMRUFTRUQiLCJtc2lzZG4iOiIiLCJzZXNzaW9uX2lkIjoiSU1TSTAwMjAyMDAwMDAwMDc4OS0xMjM0Iiwic2Vzc2lvbl9zdGFydF90aW1lIjoxNjUzNDg0MTQ0fV19fX0=", "version":"0"},
{"type": "mobilityd_ipdesc_record",
"deviceID": "IMSI001010000333333.magma.ipv4,ipv4",
"value":"eyJpcCI6IHsiYWRkcmVzcyI6ICJ3S2lBSVE9PSJ9LCAiaXBCbG9jayI6IHsibmV0QWRkcmVzcyI6ICJ3S2lBQUE9PSIsICJwcmVmaXhMZW4iOiAyNH0sICJzdGF0ZSI6ICJBTExPQ0FURUQiLCAic2lkIjogeyJpZCI6ICJJTVNJMDAxMDEwMDAwMzMzMzMzLm1hZ21hLmlwdjQsaXB2NCJ9LCAidHlwZSI6ICJJUF9QT09MIn0=", "version":"0"}
]}'
```

also works. The respective states get written into the `gateway_subscriber_states` table as above and `subscriberdb_ip_to_imsi`.





## Additional Information

- Done in pairing with @sebathomas.
- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

alexzurbonsen committed:
Add reindexing of gateway_subscriber_state to subscriberdb indexer

Modifies the existing indexer servicer by adding the subscriber
storage table to it and functionality to take states of type
"gateway_subscriber_state" (sent by sessiond) and write an entry
for each subscriber into the subscriber storage.

For that we modify the existing indexImpl function to distinguish
between the two possible types ("mobilityd_ipdesc_record" and
"gateway_subscriber_state").

Tests for the new indexer functionality are added.
A custom serde type is registered for the gateway subscriber state
and implemented in storage/subscriber_state.go.

The existing interface of the subscriber storage is changed to use
this new type (GatewaySubscriberState).

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen commented on lte/cloud/go/services/subscriberdb/servicers/protected/indexer_servicer_test.go:
```diff
@@ -96,3 +100,128 @@ func encodeIP(ip string) string {
+}
+
+
+func CreateTestGatewaySubscriberState(imsis ...string) storage.GatewaySubscriberState {
```
This is a copy and still needs a better place. One can't import functions from _test.go files.
(Tried: test_utils.go. Problem: When putting this function into a test_utils.go one has to import storage.GatewaySubscriberState and in subscriber_storage_test.go one has to import test_utils.go, which results in a circle that is not allowed.)

sebathomas replied:
I don't see how that leads to a circle. Both files import GatewaySubscriberState but that shouldn't be a problem?

alexzurbonsen replied:
I think the problem arises since the subscriber_storage_test.go file also belongs to the storage package. Then it goes like that:

package test_utils (test_utils.go) --> package storage (subscriber_state.go/GatewaySubscriberState)
package storage (subscriber_state_test.go) --> package test_utils (test_utils.go/Create...)

sebathomas replied:
Can the file test_utils also go into the package storage?

alexzurbonsen replied:
Ah, sure. That should work.

sebathomas commented on feg/cloud/go/go.mod:
```diff
@@ -94,7 +94,7 @@ require (
 	golang.org/x/sys v0.0.0-20220319134239-a9b59b0215f8 // indirect
 	golang.org/x/text v0.3.7 // indirect
-	golang.org/x/tools v0.1.8 // indirect
+	golang.org/x/tools v0.1.10 // indirect
```
Maybe all changes to `x/tools` were because of the map-merge import we had at some point that turned out not to be necessary?

alexzurbonsen replied:
I also thought so. But the import is gone and go mod tidy still returns these changes. May be it is a dependency of go-cmp.

Also the import path was "golang.org/x/exp/maps"

alexzurbonsen replied:
Theory discarded... Apparently go-cmp has no depencies outside of std lib.

sebathomas replied:
If you revert all changes to go.mod files, do they come back? I think there are only version upgrades and I don't see why they should be required here...

alexzurbonsen replied:
Indeed. I was trusting the algorithm too much ;) I thought things like that should not happen with that...

sebathomas commented on lte/cloud/go/services/subscriberdb/servicers/protected/indexer_servicer_test.go:
```diff
@@ -96,3 +100,128 @@ func encodeIP(ip string) string {
+	assert.NoError(t, err)
+	assert.True(t, cmp.Equal(*foundStates, gatewayState0))
+
+	// mixed state (two gateways, two mobilityd states)
```
Nitpick: I think it might be nice to have a separate test for the mixed state update.

sebathomas commented on lte/cloud/go/services/subscriberdb/servicers/protected/indexer_servicer.go:
```diff
@@ -59,16 +65,17 @@ type indexerServicer struct{}
 //	- an {IP -> IMSI} mapping may be stale (caller should check for staleness)
-func NewIndexerServicer() protos.IndexerServer {
-	return &indexerServicer{}
+// TODO adapt doc string
```
TODO :)

alexzurbonsen replied:
Thx!

alexzurbonsen committed:
Add "gateway_subscriber_state" annotation

To trigger the indexer also for the type gateway_subscriber_state
the annotation is added to the service registry config of
subscriberdb. k8s config is added in a separated PR.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen commented on lte/cloud/helm/lte-orc8r/examples/minikube.values.yaml:
```diff
@@ -111,7 +111,7 @@ subscriberdb:
       orc8r.io/swagger_spec: "true"
     annotations:
-      orc8r.io/state_indexer_types: "mobilityd_ipdesc_record"
+      orc8r.io/state_indexer_types: "mobilityd_ipdesc_record,gateway_subscriber_state"
```
not sure about that. I didn't touch the indentation and the other lines also have 6...
